### DISCUSSION TITLE
BRouter (offline) internal routing

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#380] [Q&A] [answered]
We can integrate BRouter (offline) inside Cruiser.

(as some users may find complicated to work with the external BRouter app)

- We will only need Cruiser for offline BRouter routing
  - No need to install the external BRouter app

- BRouter routing data downloader will work inside Cruiser

- Routing with external BRouter app will remain as option
(useful if we use BRouter with other apps)

This way will allow more intelligent downloads in the future:
download offline maps with poi along with their routing data
and download automatically the data for area of the routing.

Note: [c:geo](https://play.google.com/store/apps/details?id=cgeo.geocaching) geocaching app already integrates BRouter with nice results.

<img src="https://github.com/devemux86/cruiser/assets/3484020/09f7eb9f-4ef0-47db-a715-37f34197d6c2" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/cc264cf1-6c43-4576-a356-ee5dc4a066b5" width="50%"/>

**Comment by WolfgangZ42** [comment]
Sounds good, although the current situation is already easy to handle: just install the BRouter App, everthing else can be done by Cruiser download and setting already.

**Reply by devemux86** [reply]
Indeed, and if we use BRouter with other apps (e.g. OsmAnd, Locus), the external BRouter app is the recommended way.

We're just trying to simplify the process for non experienced users.
In addition, it will allow to unify all downloads (map + poi + routing).

**Comment by umpfmpf** [comment]
Is it possible to take over the permitted speed from brouter or does not offer it to Brouter?

**Reply by devemux86** [reply]
@umpfmpf do you mean the speed limit in the routing data?

**Reply by umpfmpf** [reply]
Yes. So that I can see the speed limit during navigation, even when I missed the calculated Route.

**Reply by devemux86** [reply]
Speed limits are available when we are on the route, as they are provided by the calculated routes.

Currently BRouter does not provide speed limits. But I may have news on that soon. :slightly_smiling_face:

**Reply by umpfmpf** [reply]
I'm so excited! The next season is just around the corner

**Comment by Ciutnik** [comment]
Perfect solution for beginners! But not only 👍

If it used somehow cleverly modified brf files containing street names (and speed limits) more users - including me - would be delighted... 🙂

**Comment by devemux86** [comment]
There will be a big performance improvement in the Cruiser code of BRouter processing.

You can expect a speedup of at least **10-20 seconds** in BRouter in the next Cruiser update.

**Reply by Ciutnik** [reply]
This is really great news! 👍 We all experience long or even longer waiting times 🙂 

Does this apply to the "external" or the "internal" BRouter or both?

**Reply by devemux86** [reply]
It will apply to both BRouter modes.

(more evident in long route calculations)

**Reply by DolomitesTom** [reply]
Will this be in a 36.bX or in the next playstore version?

**Reply by devemux86** [reply]
@DolomitesTom we will see, I am still working on several things. :slightly_smiling_face:

**Comment by devemux86** [comment]
Internal BRouter (offline) will allow us to offer more "smart" routing functions:

- When a route calculation is not possible because of missing BRouter data, a dialog will ask to check available routing data.
Then the internal BRouter download manager will open with the **needed route area** automatically selected for download.
(images 1 + 2)

- There will be a new function when long press the map that can scan and download the BRouter data for the visible map.
Then the internal BRouter download manager will open with the **visible map area** automatically selected for download.
(images 3 + 4)

|Download data for route|
|-|
|<img src="https://github.com/devemux86/cruiser/assets/3484020/0773dc83-1151-4a9b-b987-0064c49301dd" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/8b0adca3-3082-4fe9-9260-c7fde202f49e" width="50%"/>|

|Download data for map|
|-|
|<img src="https://github.com/devemux86/cruiser/assets/3484020/bfa35fcf-568b-4f43-9da3-038044e842bb" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/0fd1e668-5dc4-46a4-976d-94bd4f500221" width="50%"/>|

**Reply by devemux86** [reply]
Also as promised, another "smart" function of the internal BRouter (offline) will be available:
when we download offline maps, the BRouter data for the maps will also be **automatically** downloaded.

---

For example, when we download Austria offline map,
then Austria offline poi and Austria offline BRouter data will also be automatically downloaded.

In the end we will have Austria completely offline: map browsing, POI search, route calculation.
Everything with 1 click without leaving Cruiser environment. :slightly_smiling_face:

**Reply by kds112** [reply]
Is this download always automatic? Normally I need the map and the POIs, not routing files... Maybe introduce a new advanced setting which files are downloaded automatically?

**Reply by devemux86** [reply]
> Is this download always automatic?

@kds112 it is currently automatic if we have selected the **internal** BRouter (offline) routing.
(see the setting in the image [above](https://github.com/devemux86/cruiser/discussions/380#discussion-5864388))

Or do you mean when we select another router in Cruiser settings?

**Reply by kds112** [reply]
Ahh.. I have selected an online routing service (without using it). If the download is linked to  the routing service, everything is okay 👍

**Comment by devemux86** [comment]
BRouter (offline) internal routing is available in:
- https://github.com/devemux86/cruiser/discussions/390

With all the features mentioned above [here](https://github.com/devemux86/cruiser/discussions/380#discussioncomment-7672452) and [here](https://github.com/devemux86/cruiser/discussions/380#discussioncomment-7683029) and the _improved calculation time_.

**Comment by devemux86** [comment]
**Internal** BRouter downloader will use a blue grid color versus the **external** BRouter app with a green grid.
For better visibility and easier identification of the BRouter you are using.

Internal BRouter|External BRouter
-|-
<img src="https://github.com/devemux86/cruiser/assets/3484020/cc264cf1-6c43-4576-a356-ee5dc4a066b5"/>|<img src="https://github.com/devemux86/cruiser/assets/3484020/39f9e40f-cb6a-4a61-b5ca-960094edac6d"/>

**Comment by devemux86** [comment, marked as answer]
BRouter (offline) internal routing is available in [Cruiser 3.0.37](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
